frontend/src/app/core/optimistic-action-handler.spec.ts

```diff
@@ -12,213 +18,300 @@ class WordStateManager {
+        await actionProposalA.performAction().resolveProposal();
         expect(wordStateManager.state.word).toEqual('ac');
-        // TODO: convert from jasmine to jest
-        // expectAsync(normalProposalB).toBePending();
-        // expectAsync(optimisticProposalC).toBePending();
-        // expectAsync(normalProposalD).toBePending();
-
-        optimisticActionHandler.performAction(actionB);
-        expect(wordStateManager.state.word).toEqual('acb');
-
-        wordStateManager.respondToProposedAction({ success: true });
-        await normalProposalB;
-        expect(wordStateManager.state.word).toEqual('acb');
-        // TODO: convert from jasmine to jest
-        // expectAsync(optimisticProposalC).toBePending();
-        // expectAsync(normalProposalD).toBePending();
-
-        optimisticActionHandler.performAction(actionC);
-        expect(wordStateManager.state.word).toEqual('acbc');
-
-        wordStateManager.respondToProposedAction({ success: true });
-        await optimisticProposalC;
+
+        await actionProposalB.performAction().resolveProposal();
         expect(wordStateManager.state.word).toEqual('abc');
-        // TODO: convert from jasmine to jest
-        // expectAsync(normalProposalD).toBePending();
 
-        optimisticActionHandler.performAction(actionD);
-        expect(wordStateManager.state.word).toEqual('abcd');
+        await actionProposalC.performAction().resolveProposal();
+        expect(wordStateManager.state.word).toEqual('abc');
 
-        wordStateManager.respondToProposedAction({ success: true });
-        await normalProposalD;
+        await actionProposalD.performAction().resolveProposal();
         expect(wordStateManager.state.word).toEqual('abcd');
     });
+
+    it('handles a complicated sequence of actions', async () => {
+        const actionProposalA = proposeAction(new AddLetterAction('a'), true);
+        expect(wordStateManager.state.word).toEqual('a');
+        performAction(new AddLetterAction('b'));
+        expect(wordStateManager.state.word).toEqual('ba');
+        performAction(new AddLetterAction('c'));
+        expect(wordStateManager.state.word).toEqual('bca');
+        performAction(new AddLetterAction('c'));
+        expect(wordStateManager.state.word).toEqual('bcca');
+        const actionProposalD1 = proposeAction(new AddLetterAction('d'), false);
+        expect(wordStateManager.state.word).toEqual('bcca');
+        const actionProposalD2 = proposeAction(new AddLetterAction('d'), true);
+        expect(wordStateManager.state.word).toEqual('bccad');
+        performAction(new AddLetterAction('e'));
+        expect(wordStateManager.state.word).toEqual('bccead');
+        const performedActionProposalA = actionProposalA.performAction();
+        expect(wordStateManager.state.word).toEqual('bccead');
+        const performedActionProposalB = actionProposalD1.performAction();
+        expect(wordStateManager.state.word).toEqual('bccead');
```

Comment:
Shouldn't there be two `d`s? One from the optimistic update (`actionProposalD2`, line 261) and one from the performed action (`actionProposalD1` and `performedActionProposalB`, lines 259 and 267)?

Reply:
The first `d` is optimistic proposed and, therefore, the one in the state.  The second one only gets proposed (non-optimistic) and gets rejected in line 269. The second one never made it to the state.

Reply:
The only way to see whether one action belongs to a proposal is by comparing the sent action with the received action by value. Because D1 and D2 are identical, I assume the performAction from D1 is the performAction from D2. Therefore, there is only one `d` here.

Stuff to change: 

* [ ] Rename  `performedActionProposalB` to `performedActionProposalD1`
* [ ] Resolve  `performedActionProposalD1` at the end

Reply:
This is the same behaviour as in line 287

Reply:
Ah, I see. Fine then. You can resolve this conversation when you've included the todos.